rlbot/src/state_builder.rs

```diff
@@ -0,0 +1,534 @@
+use rlbot_flat::flat::{
+    DesiredBallState, DesiredCarState, DesiredGameState, DesiredMatchInfo, Rotator, RotatorPartial,
+    Vector3, Vector3Partial,
+};
+use std::ops::Range;
+
+/// Utility for easy construction of [DesiredGameState]s using builder patterns.
+///
+/// Example:
+/// ```example
+/// let dgs: DesiredGameState = DesiredStateBuilder::new()
+///             .car(0, |c| c
+///                 .location(Vector3::default())
+///                 .boost(100.))
+///             .ball(0, |b| b
+///                 .location_z(0.)
+///                 .velocity_z(0.)
+///             )
+///             .build();
+/// ```
+#[derive(Default, Debug, Clone, PartialEq, PartialOrd)]
+pub struct DesiredStateBuilder {
+    state: DesiredGameState,
+}
+
+#[allow(dead_code)]
+impl DesiredStateBuilder {
+    pub fn new() -> Self {
+        Self {
+            state: DesiredGameState::default(),
+        }
+    }
+
+    /// Modify the desired match info.
+    pub fn match_info(
+        mut self,
+        build: impl FnOnce(DesiredMatchInfoBuilder) -> DesiredMatchInfoBuilder,
+    ) -> Self {
+        build(DesiredMatchInfoBuilder::new(
+            &mut self.state.match_info.get_or_insert_default(),
+        ));
+        self
+    }
+
+    /// Modify the desired ball at the given index.
+    pub fn ball(
+        mut self,
+        index: usize,
+        build: impl FnOnce(DesiredBallBuilder) -> DesiredBallBuilder,
+    ) -> Self {
+        while self.state.ball_states.len() <= index {
+            self.state.ball_states.push(Default::default());
+        }
+        build(DesiredBallBuilder::new(&mut self.state.ball_states[index]));
+        self
+    }
+
+    /// Modify all desired balls.
+    pub fn all_balls(
+        mut self,
+        range: Range<usize>,
+        build: impl Fn(usize, DesiredBallBuilder) -> DesiredBallBuilder,
+    ) -> Self {
+        while self.state.ball_states.len() < range.end {
+            self.state.ball_states.push(Default::default());
+        }
+        for (i, ball) in self.state.ball_states[range].iter_mut().enumerate() {
+            build(i, DesiredBallBuilder::new(ball));
+        }
+        self
+    }
```

Comment:
Perhaps a better solution would look like this? What if you want to state-set only balls 0 and 2 for example. This applies to `all_cars` too.
```suggestion
    /// Modify all desired balls.
    pub fn all_balls(mut self, build: impl IntoIterator<Item = (usize, DesiredBallState)>) -> Self {
        for (i, ball) in build.into_iter() {
            if self.state.ball_states.len() <= i {
                self.state.ball_states.resize(i + 1, Default::default());
            }
            self.state.ball_states[i] = ball;
        }
        self
    }
```

Reply:
If you only state-set ball 0 and 2, then you should probably use the `ball` method twice instead. Changing range to `impl IntoIterator<Item = usize>` is a good idea though.

Regarding passing in DesiredXState instead of functions - If you already have an iterator of DesiredXState, then the normal API is not hard to use.

Reply:
Stuff changed in 7f7835e58694bee0ca041338e8406ef145dc07fd